subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultPluginContainer.java

```diff
@@ -111,7 +111,7 @@ public Plugin findPlugin(String id) {
     @Override
     public <P extends Plugin> P findPlugin(Class<P> type) {
         for (Plugin plugin : this) {
-            if (plugin.getClass().equals(type)) {
+            if (type.isAssignableFrom(plugin.getClass())) {
```

Comment:
Won't this break existing functionality?

Reply:
Do you have an example in mind that does not work anymore?

Reply:
One thing that certainly needs fixing is that we don't accept `null` anymore as type. Makes little sense but that was the behavior before.

Reply:
It's pretty common to see in the wild plugins which have a base class and subclasses. If you look for the common super class, then it will return the first of the plugins with that type. If you look for a concrete type it will only return the one exactly matching.

Reply:
Fair point as the base class itself may be a valid plugin.  Any thoughts on how to better handle this instead? As far as I understand, `$Instance` is just a naming convention so the plugin class we're looking for is not the enclosing class either. I don't have enough insight into `AsmBackedClassGenerator` to say what exactly we generate there and how to determine if this is just a generated proxy (except for looking at the name). Do those generated classes have something to tell is that we need to look further?

Reply:
Generated classes should implement the `org.gradle.internal.instantiation.generator.ClassGenerator.GeneratedClass` from which you can get the type of the generated class.

Reply:
Thanks! I think you meant `GeneratedSubclass`. Reworking the fix now, adding more test cases to ensure we don't break the existing contract